File: collection/Trees/BinaryTree/BinarySearchTree/TreeTraversals/TraversalsInBinarySearchTree.java
package BinaryTrees;

public class TraversalsInBinarySearchTree {

    private Node root = null;


    /** ************************** TRAVERSALS ************************** **/
    public void preOrderTraversal() {
        Node current = root;
        recursivePreOrderTraversal(current);
    }

    public void inOrderTraversal() {
        Node current = root;
        recursiveInOrderTraversal(current);
    }

    public void postOrderTraversal() {
        Node current = root;
        recursivePostOrderTraversal(current);
    }

    /** ************************** TRAVERSALS ************************** **/

    private void recursivePreOrderTraversal(Node current) {
        // Print root node
        System.out.print("[");
        System.out.print(current.value);
        System.out.print("] ");

        // Check if there is left subTree ---> Traverse left
        if(current.left!=null){
            // Traverse to left of subtree
            recursivePreOrderTraversal(current.left);
        }

        if(current.right!=null){
            // Traverse to right of subtree
            recursivePreOrderTraversal(current.right);
        }

    }

    private void recursiveInOrderTraversal(Node current) {
        // Check if there is left subTree ---> Traverse left
        if(current.left!=null){
            // Traverse to left of subtree
            recursiveInOrderTraversal(current.left);
        }

        // Print root node
        System.out.print("[");
        System.out.print(current.value);
        System.out.print("] ");

        if(current.right!=null){
            // Traverse to right of subtree
            recursiveInOrderTraversal(current.right);
        }
    }

    private void recursivePostOrderTraversal(Node current) {
        // Check if there is left subTree ---> Traverse left
        if(current.left!=null){
            // Traverse to left of subtree
            recursivePostOrderTraversal(current.left);
        }

        if(current.right!=null){
            // Traverse to right of subtree
            recursivePostOrderTraversal(current.right);
        }

        // Print root node
        System.out.print("[");
        System.out.print(current.value);
        System.out.print("] ");
    }

    /** ************************ HELPER FUNCTIONS ********************** **/
    /**
     * Inserting a Node
     */
    public void insert(int value) {

        Node current = root;

        if(current==null){
            // There are no elements inserted, so we perform the first insertion operation
            root = createNode(value);
        }else{
            Node nodeToInsert = createNode(value);
            // Iterate to the leaf node from the root where we need to perform insertion
            while (current!=null){
                // Either we insert to left of root or we insert to right of root
                if(value < current.value){
                    if(current.right==null){
                        current.left = nodeToInsert;
                        break;
                    }else{
                        current = current.left;
                    }
                }else{
                    if(current.right==null){
                        current.right = nodeToInsert;
                        break;
                    }else{
                        current = current.right;
                    }
                }
            }
        }

    }

    public Node printResult(){
        return root;
    }
    
    private Node createNode(int value) {
        return new Node(value);
    }

    private class Node {
        Node left;
        Node right;
        int value;
        public  Node(int value){
            this.value = value;
        }

        @Override
        public String toString() {
            return "Node=" + value;
        }
    }
    /** ************************ HELPER FUNCTIONS ********************** **/

}

File: collection/Trees/BinaryTree/BinarySearchTree/TreeTraversals/TraversalsInBinarySearchTree.kt
class TraversalsInBinarySearchTree {

    // Root Node
    var root : Node? = null
    // Keeping track of size
    private var size = 0



    fun preOrderTraversal() {
        var current = root
        recursivePreorderTraversal(current)
    }

    fun inOrderTraversal() {
        var current = root
        recursiveInOrderTraversal(current)
    }

    fun postorderTraversal() {
        var current = root
        recursivePostOrderTraversal(current)
    }

    private fun recursivePreorderTraversal(current: Node?) {
        if(current==null) return
        print(current.info)
        recursivePreorderTraversal(current.leftNode)
        recursivePreorderTraversal(current.rightNode)
    }

    private fun recursiveInOrderTraversal(current: Node?) {
        if(current==null) return
        recursiveInOrderTraversal(current.leftNode)
        print(current.info)
        recursiveInOrderTraversal(current.rightNode)
    }

    private fun recursivePostOrderTraversal(current: Node?) {
        if(current==null) return
        recursivePostOrderTraversal(current.leftNode)
        recursivePostOrderTraversal(current.rightNode)
        print(current.info)
    }


    /** ********************* HELPER FUNCTIONS ********************* **/
    class Node(val info : Int) {
        var leftNode : Node? = null
        var rightNode : Node? = null
        override fun toString() = "Node -> $info"
    }

    /**
     * Printing the root node
     */
    fun printNode() = root

    /**
     * Creating a new node
     */
    fun createNode(info : Int) : Node = Node(info)

    /**
     * Inserting a new element
     */
    fun insert(info : Int) {

        val newNode = Node(info)

        if(size==0){
            // If the element we are inserting is the first element, Then it becomes the root node
            root = newNode
            size+=1
        }else{
            // At-least there is one element ----> Now determine if we can add to left or if we can add to right
            var current = root

            while(current!=null){

                if(current.info==info){
                    // Don't do anything
                }else if(info<current.info){
                    if(current.leftNode==null){
                        // We reached to leaf position
                        // Insert to the left of the tree
                        current.leftNode = newNode
                        size += 1
                        break
                    }else{
                        // Traverse to left of the current node
                        current = current.leftNode
                    }
                }else{
                    if(current.rightNode==null){
                        // We reached to leaf position
                        // Insert to the right of the tree
                        current.rightNode = newNode
                        size += 1
                        break
                    }else{
                        // Traverse to right of the current node
                        current = current.rightNode
                    }
                }

            }
        }



    }
    /** ********************* HELPER FUNCTIONS ********************* **/


}